### PR TITLE
storage-client: enable compaction of progress topics

### PR DESCRIPTION
**Pull request by benesch**
When Materialize creates a progress topic, enable key-based compaction on the topic so that progress records do not pile up forever. Also add some documentation that points users in the right direction if they choose to manage the creation of topics themselves.

### Motivation

  * This PR adds a feature that has not yet been specified.

### Tips for reviewer

Noticed somewhere in the flurry of recent sink work that we were failing to do this.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Automatically enable compaction when creating the progress topic for a Kafka sink. **Warning:** Versions of Redpanda before v22.3 do not support using compaction for Materialize's progress topics. You will need to manually create the connection's progress topic with compaction disabled to use sinks with these versions of Redpanda.

### CODE REVIEW

File: src/storage-client/src/sink.rs
Comment (benesch) [inline src/storage-client/src/sink.rs:368]: Note this is not actually a change in behavior because `REPLICATION FACTOR` is an unstable option that is forced to -1 in stable mode. But, were that option stable, it seemed like it'd be very wrong for the progress topic to get created with the replication factor of whichever sink happened to win the race to create the progress topic.

File: src/storage-client/src/sink.rs
Comment (benesch) [inline src/storage-client/src/sink.rs:373]: This line is the only true behavior change of this PR.

File: src/storage-client/src/sink.rs
Comment (benesch) [inline src/storage-client/src/sink.rs:372]: Explicitly disabling retention to override whatever broker default retention policy might exist seemed worthwhile.

File: doc/user/content/sql/create-sink/kafka.md
Comment (morsapaes) [inline doc/user/content/sql/create-sink/kafka.md:361]:
```suggestion
If the connection's [progress topic](#exactly-once-processing) does
not exist, Materialize will also attempt to create it when the sink starts.
The progress topic will be configured with a single partition, the
broker's default replication factor, and compaction enabled.
```

File: doc/user/content/sql/create-sink/kafka.md
Comment (morsapaes) [inline doc/user/content/sql/create-sink/kafka.md:377]:
```suggestion
| Data topic     | Partition count     | Your choice, based on your performance and ordering requirements.
| Data topic     | Replication factor  | Your choice, based on your durability requirements.
| Data topic     | Compaction          | Your choice, based on your downstream applications' requirements. If using the [Upsert envelope](#upsert), we recommend enabling compaction.
| Data topic     | Retention           | Your choice, based on your downstream applications' requirements.
| Progress topic | Partition count     | **Must be set to 1**. Using multiple partitions can cause Materialize to violate its [exactly-once guarantees](#exactly-once-processing).
| Progress topic | Replication factor  | Your choice, based on your durability requirements.
| Progress topic | Compaction          | We recommend enabling compaction to avoid accumulating unbounded state. Disabling compaction may cause performance issues, but will not cause correctness issues.
| Progress topic | Retention           | **Must be disabled**. Enabling retention can cause Materialize to violate its [exactly-once guarantees](#exactly-once-processing).
```

File: doc/user/content/sql/create-sink/kafka.md
Comment (morsapaes) [inline doc/user/content/sql/create-sink/kafka.md:363]:
```suggestion
To customize the topics' configurations, manually create the topics in Kafka
```

File: doc/user/content/sql/create-sink/kafka.md
Comment (morsapaes) [inline doc/user/content/sql/create-sink/kafka.md:364]:
```suggestion
with the desired configuration (e.g., using the [`kafka-topics.sh`]
```

**Review by morsapaes — CHANGES_REQUESTED** [review]

**Comment by bkirwi** [comment]
Had a look at the cloudtest output:

> default/redpanda-7b846d8999-59rql[redpanda]: ERROR 2023-11-26 23:30:35,557 [shard 0] cluster - rm_stm.cc:228 - Can't process a transactional request to {kafka/_materialize-progress-cloudtest-test-00000000-0000-0000-0000-000000000000-0-u2/0}. Compacted topic doesn't support transactional processing.

Which is weird because [these docs](https://docs.redpanda.com/current/develop/transactions/#transactions-with-compacted-segments) say:

> Transactions are supported on topics with compaction configured. When [log_cleanup_policy](https://docs.redpanda.com/current/reference/cluster-properties/#log_cleanup_policy) or [cleanup.policy](https://docs.redpanda.com/current/reference/topic-properties/#cleanuppolicy) are set to either compact or compact,delete, the compaction process removes the aborted transaction’s data and all transactional control markers from the log. The resulting compacted segment contains only committed data batches (and potentially harmless gaps in the offsets due to skipped batches).

Maybe we need to upgrade redpanda in cloudtest?

**Comment by benesch** [comment]
> Maybe we need to upgrade redpanda in cloudtest?

Oh, brilliant, I hadn't dug deep enough to find that. Looks like absolutely yes. Looks like txns on compacted topics were enabled around v22.3 (https://github.com/redpanda-data/redpanda/commit/a4c9521e525f3881d936470a8192be08c3bb2491), and cloudtest is still on the relatively ancient v21.11.13.

**Comment by benesch** [comment]
Updated the Redpanda version in cloudtest and all seems to be well now. I split out the docs into their own PR (https://github.com/MaterializeInc/materialize/pull/23601) to not block this on @morsapaes's feedback. So assuming this goes green, just awaiting your approval, @bkirwi!

File: misc/python/materialize/cloudtest/k8s/redpanda.py
Comment (nrainer-materialize) [inline misc/python/materialize/cloudtest/k8s/redpanda.py:37]: Perfect, this version is in line with https://github.com/MaterializeInc/materialize/pull/23567/files.

**Comment by nrainer-materialize** [comment]
I triggered a nightly build: https://buildkite.com/materialize/nightlies/builds/5453

File: src/storage-client/src/sink.rs
Comment (bkirwi) [inline src/storage-client/src/sink.rs:241]: 💭 This is correct, but I'd also be tempted to ban mixing these... I've never seen a case where it's actually what you want. (vs. https://kafka.apache.org/documentation/#topicconfigs_max.compaction.lag.ms which has some advantages.)

**Review by bkirwi — APPROVED** [review]
This is good!

The nightly failures seem to be half network flakes, and one test that [explicitly checks old redpanda versions](https://github.com/MaterializeInc/materialize/blob/24cd2b53213a93ea3f9d313c1be7a125a3d8f7de/test/kafka-matrix/mzcompose.py#L19). If you're comfortable dropping support for aging redpandas then that should be an easy fix!

File: src/storage-client/src/sink.rs
Comment (benesch) [reply src/storage-client/src/sink.rs]: Fair point! I was trying to keep the utility function as general purpose as possible, but happy to make its signature more prescriptive and force the caller to choose between retention and compaction. Will push up a revision to handle that.

Comment (benesch) [reply src/storage-client/src/sink.rs]: Done.

**Review by bkirwi — APPROVED** [review]
Looks great!

**Comment by benesch** [comment]
Thanks for the reviews!

**Pull request merged**
